webrender/src/render_backend.rs

```diff
@@ -510,20 +549,82 @@ impl RenderBackend {
         loop {
             profile_scope!("handle_msg");
 
-            let msg = match self.api_rx.recv() {
+            while let Ok(msg) = self.scene_rx.try_recv() {
+                match msg {
+                    SceneBuilderMsg::Transaction {
+                        document_id,
+                        mut built_scene,
+                        resource_updates,
+                        frame_ops,
+                        render,
+                    } => {
+                        if let Some(doc) = self.documents.get_mut(&document_id) {
+                            if let Some(mut built_scene) = built_scene.take() {
+                                doc.current.scene = built_scene.scene;
+                                doc.frame_builder = Some(built_scene.frame_builder);
+                                doc.current.removed_pipelines.extend(built_scene.removed_pipelines.drain(..));
+                                doc.frame_ctx.new_async_scene_ready(
+                                    built_scene.clip_scroll_tree,
+                                    built_scene.pipeline_epoch_map,
+                                );
+                                doc.render_on_hittest = true;
+                            }
+                        } else {
+                            // The document was removed while we were building it, skip it.
+                            // TODO: we might want to just ensure that removed documents are
```

Comment:
It seems like there might always be the possibility for a race condition here even if the removed documents are forwarded to the scene building thread?